src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -130,104 +156,133 @@ public static uint GetResponseHeaderNumberInfo(SafeWinHttpHandle requestHandle,
             return result;
         }
 
-        private static string GetResponseHeaderStringInfo(SafeWinHttpHandle requestHandle, uint infoLevel)
+        private static char[] CreateResponseHeaderInfoBuffer(SafeWinHttpHandle requestHandle, uint infoLevel)
         {
-            uint bytesNeeded = 0;
-            bool results = false;
+            uint bufferLengthInBytes = 0;
 
-            // Call WinHttpQueryHeaders once to obtain the size of the buffer needed.  The size is returned in
+            // Call WinHttpQueryHeaders to obtain the size of the buffer needed. The size is returned in
             // bytes but the API actually returns Unicode characters.
             if (!Interop.WinHttp.WinHttpQueryHeaders(
                 requestHandle,
                 infoLevel,
                 Interop.WinHttp.WINHTTP_HEADER_NAME_BY_INDEX,
-                null,
-                ref bytesNeeded,
+                IntPtr.Zero,
+                ref bufferLengthInBytes,
                 IntPtr.Zero))
             {
                 int lastError = Marshal.GetLastWin32Error();
-                if (lastError == Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND)
+
+                Debug.Assert(lastError != Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND);
+
+                if (lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER)
                 {
-                    return null;
+                    throw WinHttpException.CreateExceptionUsingError(lastError);
                 }
+            }
 
-                if (lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER)
+            return new char[bufferLengthInBytes / 2];
+        }
+
+        private static int GetResponseHeaderInfo(SafeWinHttpHandle requestHandle, uint infoLevel, char[] buffer)
+        {
+            Debug.Assert(buffer != null, "buffer must not be null.");
+            Debug.Assert(buffer.Length > 0, "buffer must not be empty.");
+
+            uint bufferLengthInBytes = (uint)buffer.Length * 2;
+
+            GCHandle pinnedHandle = GCHandle.Alloc(buffer, GCHandleType.Pinned);
```

Comment:
Switched to using `fixed`. The reason I used `GCHandle` here was because this is how it was being done elsewhere in WinHttpHandler. (When I originally submitted this PR, the code in WinHttpHandler looked like [this](https://github.com/dotnet/corefx/blob/a75ec027daa53114c788464878e2947ae7d5a65e/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs#L100), for example. It has since changed a bit: it's still using `GCHandle` but is doing some caching of the handle now).